velox/substrait/SubstraitToVeloxPlan.cpp

```diff
@@ -0,0 +1,504 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "SubstraitToVeloxPlan.h"
+#include "TypeUtils.h"
+
+namespace facebook::velox::substrait {
+
+std::shared_ptr<const core::PlanNode> SubstraitVeloxPlanConverter::toVeloxPlan(
+    const ::substrait::AggregateRel& sAgg) {
+  std::shared_ptr<const core::PlanNode> childNode;
+  if (sAgg.has_input()) {
+    childNode = toVeloxPlan(sAgg.input());
+  } else {
+    VELOX_FAIL("Child Rel is expected in AggregateRel.");
+  }
+  auto inputTypes = childNode->outputType();
+  // Construct Velox grouping expressions.
+  std::vector<std::shared_ptr<const core::FieldAccessTypedExpr>>
+      veloxGroupingExprs;
+  const auto& groupings = sAgg.groupings();
+  int inputPlanNodeId = planNodeId_ - 1;
+  // The index of output column.
+  int outIdx = 0;
+  for (const auto& grouping : groupings) {
+    auto groupingExprs = grouping.grouping_expressions();
+    for (const auto& groupingExpr : groupingExprs) {
+      auto fieldExpr =
+          exprConverter_->toVeloxExpr(groupingExpr, inputPlanNodeId);
+      // Velox's groupings are limited to be Field, and pre-projection for
+      // grouping cols is not supported.
+      auto typedFieldExpr =
+          std::dynamic_pointer_cast<const core::FieldAccessTypedExpr>(
+              fieldExpr);
+      veloxGroupingExprs.emplace_back(typedFieldExpr);
+      outIdx += 1;
+    }
+  }
+  // Parse measures to get Aggregation phase and expressions.
+  bool phaseInited = false;
+  core::AggregationNode::Step aggStep;
+  // Project expressions are used to conduct a pre-projection before
+  // Aggregation if needed.
+  std::vector<std::shared_ptr<const core::ITypedExpr>> projectExprs;
+  std::vector<std::string> projectOutNames;
+  std::vector<std::shared_ptr<const core::CallTypedExpr>> aggExprs;
+  aggExprs.reserve(sAgg.measures().size());
+  // Construct Velox Aggregate expressions.
```

Comment:
nit: maybe use a blank line to separate code blocks?

Reply:
Fixed.